### ISSUE TITLE
Design Order Online page (This task will track progress on the online ordering)

### ISSUE THREAD

**Issue by NullOperator404** [NullOperator404/Developer-Diary#44]
Preliminary layout work of the online ordering page(s) is underway. I've decided to base the ordering process loosely on that of the Dickies BBQ page. This part of the website will be the most labor intensive, as it requires extensive Javascript to handle all of the functions of the ordering pages. Based on what I've seen on Dickie's, the P.B's ordering process will look something like this:

1. User clicks "Order Online" 

2. Upon arriving on the Order Online page, the user will be prompted to select which location they want to pick up from (Butler or Huntsville). Their selection here will be stored so the checkout process will know which location to send the order to. 

3. After selecting a location, a block version of the menu will appear. The default view will be a list of all menu items, but a filter option at the top of the screen will allow users to refine results to a specific set of items, like sandwiches. 

4. Clicking one of the blocks will then open a "description screen" with an image of the item, a more detailed description of the item, and options to select sides, drinks, add-on condiments, and of course, a button to add the selection to the shopping cart. 

5. The user will then have the option to make additional selections, or to go to the shopping cart to proceed with checkout. 

Much of this application will mimic the design lab I built for my own website, requiring two major things: 

1. The "description page" will be a "dummy page" with placeholders for the text and images; these would be obviously populated based on what the user has selected on the main page. That said, a JSON file will need to be written containing all this pertinent data. Things that will need to be populated are: The product image, the product title, the product description, product price, sides, drinks, and add-ons. Some items, such as "PB's Plates," come with a side, but the "Tasty Sandwiches" don't. This will need to be taken into account when building the JSON file. 

2. The shopping cart will need to be an array capable of storing the user's final selections so they can be outputted on the checkout screen and then sent to Clover. A JS system will be need to be built that ensures the shopping cart data is updated in case the user changes their mind and deletes items, changes sides, drinks, condiments, etc. The shopping cart will also need to keep track of the order total, pre-tax and post-tax, before handing off to Clover. 

Lastly, perhaps the biggest hurdle is going to be connecting these two applications. The POS system at the restaurant runs on Clover, and all orders placed at the front counter and the drive-thru are displayed on the Clover monitors. The website will need to handle accepting payment via Clover, which isn't too big of an issue; I've already been in touch with Clover about this and it operates the same as PayPal, using a token to pass credit card data to the server. The real kicker is going to be passing the ORDER DATA to the Clover system so that it will display in the restaurant for the employees. 

Monday, I will be calling Clover again to inquire about this further. In the meantime, I'll be researching it myself.

**Comment by NullOperator404** [comment]
UPDATE 11/15/2024

The initial screen for Ordering Online is almost finished - I just need to write the Javascript that will save which location the user has selected. A CSS animation will slide the current screen out-of-frame to the left and the next screen in from the right, which will be the main menu screen. This will be the largest of the three screens to build. 

TOTAL TIME IN PROJECT SO FAR: 16 hours, 3 minutes

**Comment by NullOperator404** [comment]
UPDATE 11/18/2024

Considerable progress was made tonight on the Online Ordering interface. I'll break things down below:

LOCATION SELECTION: 
This part of the interface is finished. When the User clicks one of the two locations, the "Locations" screen slides out of frame with a CSS animation and the Main Menu screen slides into frame, creating a seamless transition from one screen to the next. In addition, a "Back" button was added to the main menu screen, which reverses the process and allows the User to return to the first screen where they can change the pick-up location if they wish. 

Additionally, the "Location" header on the Main Menu screen now updates depending on which location was selected. 

PRODUCT JSON DATABASE:
Since the main menu screen will be dynamic in that the User will be able to filter menu items, and since shopping cart data will need to save things like item title, description, price, etc, this necessitated the creation of a JSON file that will contain all pertinent data for each menu item. Tonight, I started building the foundation for the JSON file using the first 8 or 9 items from the menu. Each object in the JSON file will contain data for the item's CATEGORY, TITLE, DESCRIPTION, PRICE, an IMAGE, available SIDES, and DRINKS. Additionally, an eighth entry, called "SIDEINCLUDED," will have a value of either "Yes" or "No," which will then tell JavaScript whether or not to deduct the cost of the first selected side. 

DYNAMIC MENU GENERATION:
Using the created JSON file, I then setup a Fetch command that reads the JSON file and then populates the menu screen with the retrieved data. This function also handles building out the individual "blocks" for each item to be displayed in, which in turn, will act as buttons -- when the User clicks and item, they will be moved to the third screen (in the same manner they were moved to the menu screen), which will allow them to customize their item, add sides, drinks, and of course, add the selection to their shopping cart. 

MINOR ISSUES:
Naturally, when building out a large JS application like this, there are always going to be unforeseen bugs. At present, the only two minor issues are:

1. If the User selects a location and visits the Menu screen, the HEIGHT of the webpage is increased due to the menu being populated. Since the "Menu" and "Select Location" screens share the same page space, the "Select Location" screen will have a ton of empty space at the bottom (caused by the out-of-view menu screen). This can be rectified by clearing the generated menu HTML.

2. The Fetch command is reading the JSON file in reverse order. That is, what should be the first item on the menu is being displayed last. I'm not entirely sure why this is occurring, as the `insertAdjacentHTML` function that handles building an element is set to `afterend`, meaning JavaScript _should_ be inserting each additional element AFTER the previous one it built. A possible solution would be to rewrite the JSON file in reverse order. 

ADDENDUM: It occurred to me that the reason #2 is happening is because JavaScript is being instructed to insert new elements after the _location element,_ not after _the last element it generated._ So in essence, JS is wedging new elements between the location element and the previously generated element on each iteration. 

TOTAL TIME IN PROJECT SO FAR: 19 hours, 3 minutes

**Comment by NullOperator404** [comment]
UPDATE 11/20/2024 - 10PM

All of the issues listed in my previous entry have been resolved. Solutions below:

1. For this, it occurred to me that the menu item space (populated by JS) just needed it's own independent `<div>` container so that JS's operations would be confined to this element, thus negating the need to get the previous element involved. After adding the following HTML:

`<div id="menuSpace"></div>`

A line of code was then added to the `backUp()` command which clears any HTML elements inside the new `<div>`:

`menuSpace.innerHTML = "";`

A `setTimeout()` function was needed to delay the clearing for half a second to allow for the "slide out of frame" CSS animation:

```
//WAIT UNTIL SCREEN IS OUT-OF-FRAME, THEN CLEAR MENU SCREEN
    setTimeout(clear, 500);
        function clear() {
            menuSpace.innerHTML = "";
        }
```

2. The second problem of the JSON file being read in reverse order was also solved with the solution implemented in problem 1. 

I refactored the `insertAdjacentHTML()` function in the Fetch command to operate as `beforeend` rather than `afterend`. This way, each iteration through the database generates an HTML element AFTER the previous one generated (inside the `<div>` container. This corrected the listing order; items now display in the correct order. 

ADDITIONAL PROBLEM RESOLVED:

Another problem I neglected to mention yesterday was that I wanted the menu items to be clustered by Category, separated by a HTML banner element with the title of the current category. Until now, the program would just output all of the JSON data one after another without categorizing them. 

For this reason, a check system was implemented in the Fetch command that checks the `Category` of the entry currently being read. It then compares this to another var, `previousCat`, which contains the category of the _previous_ entry read. If they don't match, JS generates a new banner element with the new category text (pulled from the category it just read). If they do match, it ignores the banner code and continues generating a new menu item:

```
function populate(product) {
                        category = product.Category;
                        //DOES A NEW HEADER NEED TO BE BUILT?
                            if(category !== previousCat) {
                                menuSpace.insertAdjacentHTML("beforeend", "<div class='onlineMenuHeader'>" + category + "</div>");
                            }

                            //BUILD CONTAINER
                            menuSpace.insertAdjacentHTML("beforeend", "<div class='menuItemContainer'><img src='graphics/thumbs/something.jpg' class='thumbnail'><h3 class='itemHeader'>" + product.Header + "</h3><div class='description'>" + product.Description + "</div><span class='price'>$" + product.Price + "</span></div>");
                            //SAVE CURRENT CATEGORY TO VAR
                            previousCat = category;

                        }
```

With these problems now solved, I can now focus on completing the JSON file, and then working on building the "Item Details" screen where Users will be able to customize their order and add items to their shopping cart.

**Comment by NullOperator404** [comment]
UPDATE 11/20/2024 10:53 PM

The JSON file is now finished, though I may need to make additions later once I get into the shopping cart portion; the data being transferred to the cart, outputted on the checkout screen, and then of course passed to Clover is all pulled from the JSON data. 

I'm now going to work on implementing a "filter" system for the menu screen so that Users can refine results.

**Comment by NullOperator404** [comment]
UPDATE 11/21/2024 12:57 AM

HTML and CSS work for the filter system is complete. 

I've also got the bulk of the functionality done in JavaScript. A few minor additions need to be made, but the system works as intended. I've also included a "Clear filters" button which will revert the screen back to it's default configuration. 

Minor issues that need to be addressed:
1. Separate `if` statements need to be written that will handle the conditions of "Chicken" and "Combos". 

2. The filter buttons remain inactive after the "Clear filters" button has been clicked. A block of code will need to be written in the `removeFilters()` function that reverts all buttons to their default configuration. These commands also need to be added to the `backUp()` function, since it also reverts the menu screen to it's default configuration. 

3. The system needs to be debugged -- I have noticed a specific bug where if the filters are clicked in the following sequence; "Wings > BBQ", the "Wings" banner will disappear until another filter is clicked, like "Kid's Meal." This is probably something to do with the system comparing the new Category with the previous one. I'll look into it more tomorrow. 

Once the filter system is finished, I'll go about adding photos for each of the items. Then I'll be ready to start building the Item screen where Users will customize their order and add it to the shopping cart. 

TOTAL TIME IN PROJECT SO FAR: 22 hrs, 21 mins

**Comment by NullOperator404** [comment]
UPDATE 11/26/2024 

1. The separate `if` statements have been added to the filter system and now work as intended. 

2. Code has been added to the `removeFilters()` function that resets the filter buttons, but JS isn't restoring the default `;hover` styling associated with them in their default configuration, despite the `pointerEvents` property being reset to "auto." I will need to look into this further. 

3. A thorough debugging is still in order for the filter system, trying the filter buttons in different combinations to ensure the headers are displaying properly. At the moment, only the "Wings > BBQ" sequence manifest a bug in which the "Wings" banner disappears from the top of the screen after clicking "BBQ." 

Tomorrow I will debug the system and address the `:hover` issue, then probably move on to building the Item screen; for now I'll just use placeholder boxes where the images will display, as gathering the images is a nothing burger task (if you'll pardon the pun). 

One other thing on the main menu screen: I need to get in touch with management at the restaurant (preferably Denise) and have them describe how each item is served (toppings, condiments, etc). This descriptive information must then be added to the JSON file.

After the Item screen, images, and JSON files have been finalized, I'll be ready to start working on connecting the site to Clover. 

One final note: 
One issue occurred to me a few days ago regarding the menu screen -- each item listed would need a unique ID so JS could handle `onClick` functions (similar to elements in the Design Lab for my own website). Since the menu elements are generated dynamically, it is not possible to use EventListeners, because at run time the elements do not yet exist in the DOM, and thus JS has nothing to assign the Event Listeners to. 

In the Design Lab, I worked around this by having JS assign a string of characters as the prefix to an ID for a generated element, like "box", which would then be concatenated with an integer. For each element generated, the integer increments by 1, so IDs are assigned as "box1", "box2", "box3" and so on. Mind you, this example is just for illustrative purposes; the actual strings used are more specific and clearly identify what the elements are or what their purpose is.

However a few nights ago I realized that for this application, it wouldn't be necessary to use this method, as the IDs for each menu item could be stored in the JSON file, and upon generation of an element, JS can just assign an ID by pulling from the JSON file. I'm actually considering refactoring some of the code from my Design Lab using this method, but I'm not sure it would be applicable without delving in further. 

This method should simplify communication between the menu screen and the item screen, as a `getAttribute('id')` function will capture the ID of the clicked item, thus providing a reference so the Item screen can locate that item's entry in JSON and copy the data to the Shopping Cart array for use at checkout. 

TOTAL TIME IN PROJECT SO FAR: 23 hrs, 21 mins

**Comment by NullOperator404** [comment]
UPDATE 11/26/2024 7:32 PM

The debugging process is underway and as expected, it's pretty buggy. 

1. The "sandwiches" filter did not work due to a text mismatch. Synchronizing the string value in the code and in the JSON file fixed this issue.

2. The issue of the "BBQ" banner vanishing when a second filter is clicked is not exclusive to "BBQ > Wings," but rather occurs no matter which two filters are clicked. That is, when the second filter is clicked, the banner at the top of the first group always vanishes, and I now know why. 

On the first `for` loop iteration, there is only the one filter saved to the `filterArray[]`, so only 1 loop is performed. Much like the `Fetch` block of code, the function first checks to see if a banner needs to be generated by comparing the value of `currentCat` to `previousCat`. Since `previousCat` has a value of null in this instance, the condition statement `if(currentCat !== previousCat)` evaluates to true, thus triggering the creation of a banner element. At the end of this block of code, the function updates the value of `previousCat` to that of `currentCat,` (because the next item in the list is of the same category, and thus no new banner is needed). So the `if` statement evaluates to false (the two vars are equal), and JS ignores this bit of code. 

However, when the user clicks a second filter ("Wings," for example), it's saved to the `filterArray[]`, making the total number of filters 2. This means 2 loops will be performed - one for "BBQ," and one for "Wings." The problem is that each time the user clicks a filter, the function clears the existing HTML from the menu screen so the new filtered results can be outputted to the screen. This is when the top banner initially gets deleted. 

The problem manifests when the loops are performed for the second filter. The first filter click ends by synchronizing the two category vars `currentCat` and `previousCat`, so when the 2nd loop is performed, the first check to see if a banner is needed (for "BBQ") inadvertently evaluates to FALSE (because the two vars are equal), causing JS to ignore reproducing the banner. 

The "Wings" banner remains because the check statement works properly in that instance; the category has changed to "Wings," and thus no longer equals the value of `previousCat`. 

As I write this, I think I may have come up with a simple solution...

At the end of the function, a statement can be inserted that resets the value of `previousCat` to `null`, thus causing the function to run every time as though it's it's _first time._ (`previousCat`'s value is "undefined" by default.

**Comment by NullOperator404** [comment]
UPDATE 11/26/2024 7:56 PM

Success! Adding the following line code at the end of the function fixed the problem:

`previousCat = null;`

I'm now going to repeat the debug process to ensure no other errors exist, and if not I'll move on to addressing the `:hover` property staying turned off.

**Comment by NullOperator404** [comment]
UPDATE 11/26/2024 8:04 PM

A quick look on StackOverflow got a resolution to the `:hover` issue, and I feel like such a bonehead for not spotting this myself. 

When the clear filters button is clicked, JS updates the styling of the buttons to their default configuration. The problem - which I forgot - is that JS writes _inline styles,_ not the styles defined in the external stylesheet -- and inline styles supersede external styles. 

However, applying the `!important` property to each of the external styles for the `:hover` event resolved the issue. 

This concludes the debugging of the filter system, so I'm now going to work on designing the item screen.

**Comment by NullOperator404** [comment]
UPDATE 11/26/2024 9:08 PM

HTML framework and most of the CSS styling is done for the item screen. A few more things to do and I'll be ready to start working on JS functionality. 

TOTAL TIME IN PROJECT SO FAR: 25 hours, 21 mins

**Comment by NullOperator404** [comment]
UPDATE 12/16/2024

Tonight I began work on the HTML & CSS for the hidden blocks that will display the list of Sides, Drinks, and Dipping Sauces. 

This took much longer than I anticipated, as I realized the hide/reveal function of these blocks relies on their CSS positions being `relative` rather than `absolute`. When I initially styled the header bars, I set their positions to `absolute` so I could be given greater control over their placement in the DOM. Being positioned absolutely means the header bars beneath the one being clicked, like "Sides," would not drop down out of the way to reveal the options menu -- this only works if they are positioned RELATIVELY. 

So before I could setup the checkboxes & labels, I had to first refactor all of the CSS for the header bars so that they are all positioned relatively. This required a little additional HTML to divide the screen into two columns - the left for the image, and the right for the buttons. 

I will need to refactor the HTML further to customize the styling of the checkboxes, since CSS can't style them directly. 

This screen will require a ton of JavaScript, so I may have underestimated the complexity of this screen in my previous comments; not only will JS need to toggle the options menus, but it will also need to keep track of how many options have been checked, which ones have been checked, etc. 

TOTAL TIME IN PROJECT SO FAR: 26 hours, 46 mins

**Comment by NullOperator404** [comment]
UPDATE 12/18/2024

Tonight I finished refactoring the HTML & CSS to customize the checkboxes. With that completed, I was ready to write a JS function to handle the opening & closing of these hidden menus. When the user clicks one of the banners ("Sides," "Drinks" or "Dipping Sauces,"), a sub-menu is revealed showing each of the options accompanied with a checkbox. The user will be able to make as many selections as they wish, but considerable JS will be needed to handle various conditions, such as not upcharging for 2 sides if they're included in a meal. 

The hide/display function is working, but there are a few minor bugs pertaining to CSS's layering; the revealed sub-menu sits on top of the banner (in terms of layers) and thus inadvertently causes the banner to become un-clickable once the menu is open. (Clicking the banner if the menu is already open closes the menu).

Next tasks: 
1. Write a JS function that will move the user to the Item Screen. When the user clicks one of the options on the main menu screen, JS will need to populate the Item screen by retrieving data from the JSON file. This will allow me to toggle the screen back and forth, as at present it's blocking the other screens -- I needed it visible at all times while building out the HTML & CSS. 

2. Design a JS function (or functions) that will update the price total depending on selections the user makes in the submenus. 

TOTAL TIME IN PROJECT SO FAR: 28 hours, 49 mins

**Comment by NullOperator404** [comment]
UPDATE 12/19/2024

Tonight saw considerable progress with the ordering screen. Referring to my previous entry, tasks 1 & 2 have been completed. All that's needed now is to setup a function for the order screen's "back" button, which will clear the HTML from the order screen and return the user to the main menu screen. 

ITEM DISPLAY DATA:
The screen now dynamically updates depending on which selection the user has made on the main menu by retrieving data from the JSON file. Item header, description, price, and whether sides are included are all updated when the user makes a selection from the main menu. 

To allow each main menu item to have a unique ID, I added a object-value pair to each JSON entry so that each item has a pre-assigned ID. This data is then inserted into the generated HTML when the page loads, thus automatically assigning each item box an ID to be used in JS. **_The JSON file still needs to be updated with image data for each item, and the object/value pairs for "Sides" and "Drinks" can be removed, as these items ALWAYS appear on the order screen._** 

QUANTITY BUTTONS:
Additionally, I built & styled a quantity counter for the order screen. Users can now click the "-" or "+" buttons to decrease/increase the quantity of the selected item. Doing also also adjusts the total indicated price, which is now displayed next to "Add to Order" on that button. **_I still need to setup an additional function for the input field itself, so that if the user chooses to manually type in a value instead of using the buttons, the price will update accordingly. Also, a check system needs to be implemented that ensures the user inputs only numbers._** 

SIDE SELECTION:
I began work on the function that handles the selection of sides. Several conditions had to be accounted for, including whether or not sides were included with the item, how many sides had already been selected (if any), and whether or not the click being performed is "checking" or "unchecking" the box. Again, price is updated accordingly -- at present I have it set to +$3.00, but I will later need to expand the system so that users can select between the Regular or Pint size (the latter sells for $5.00 ea). 

This part still needs a little debug work, as the de-select side option is not correcting the price like it should. 

Once I get these minor bugs worked out and addressed the issues appearing bold italics, I will move on to the JS that handles the selection of drinks - which will be much more straightforward; the combos at PBs don't include drinks, but they do offer 2 sizes. 

I spoke with Denise today and she informed me that they do not offer any "Add-ons" to items like pickles, so the "Add-Ons" bar can be removed from the order screen. 

After that, I _think_ I'll be ready to start working on a function for the "Add to Order" button, which will store the order data in an object array. What happens then is up in the air --- do I return the user to the main menu screen? Or should I keep them on the existing screen? Or should a modal pop up? I'll need to sleep on this (and probably see how the Dickies site responds at this stage).

TOTAL TIME IN PROJECT SO FAR: 30 hours, 44 mins

**Comment by NullOperator404** [comment]
UPDATE 12/27/2024

I'm taking advantage of the Xmas holidays to work on the project some more. 

Tonight, I was able to resolve the issue of the inconsistent price outputs when selecting/deselecting sides. Additionally, I've written the code so that calculating the price is uniform for both adjusting quantity or selecting/deselecting sides. This logic will need to be applied to the Drinks selections, as well. Basically, a var `currentTotal` holds the total price of 1 ITEM + any sides. The quantity adjustment functions now multiply this var by `quantity` to get a total price for output on the DOM. 

Likewise, the Sides selectors simply add/subtract $3.00 from `currentTotal`, and the var is then multiplied by `quantity` to get the output amount. 

This ensures that if a user selects sides and then increases the quantity of the ITEM, then the sides are increased as well. (i.e. 1 item + 2 sides, increased by 1 becomes 2 items with 4 sides; with the sides simply being doubled). 

I've also setup a function tonight to handle updating the price should the user manually input numbers into the box. However, a check still needs to be added to ensure they ONLY input numbers. 

**CURRENT TASKS THAT NEED TO BE ADDRESSED NEXT:**
1. "Back" button needs functionality so that it returns the user to the main menu screen (basically reversing the steps of the `openOrderScreen()` function.

2. A function similar to the one handling SIDES needs to be setup for DRINKS. 

3. "Add-ons" header bar needs to be removed from HTML; not needed. 

4. JSON file needs to be updated -- remove "Sides" and "Drinks" object/value pairs; not needed in JSON file. 

5. Setup a function (or add to existing functions) allowing JS to store order info to an object array. The array needs to include ITEM TITLE, PRICE, QUANTITY, SIDES, DRINKS, DIPPING SAUCE(S), SPECIAL INSTRUCTIONS, etc. More data may need to be added once I begin building the function. The data stored in this array will be used not only for the checkout screen, but also will be used when the order is sent to Clover. 

6. Build a checkout screen that imports data from the array referenced in number 5,  and link to Clover checkout process. Rather than do another "slide" transition, this should be handled on an independent screen for security purposes -- so the array data will need to be transferred to a COOKIE on the user's device so it can be imported on the new page. 

TOTAL TIME IN PROJECT SO FAR: 32 hours, 24 mins

**Comment by NullOperator404** [comment]
UPDATE 12/28/2024

More progress tonight, and I'm getting near the end of the Order Online part of the project. Tasks 1 though 4 listed above have been completed, and I've begun scribbling down some notes for task 5. Below I'll go into further detail:

One of my main goals with this order screen was to avoid the hurdles of user indecisiveness that I had to allow for in my Design Lab project; that is, I didn't want any data being sent to the shopping cart array until the user is _absolutely sure_ of their order. Otherwise, I would need to devise systems of tracking how many sides/drinks/sauces a user has clicked, the order in which they clicked them, etc so that in case they changed their mind and de-selected one, the appropriate selection could be removed from the proper index in an array storing their selections. 

Instead, it's going to be much simpler to just have the "Add to Cart" button perform a check of all the Sides, Drinks, & Sauces to see which ones are selected. If one is, THEN it can push that data into an array in the cart. It's extremely unlikely that once a user clicks "Add to Cart," they'll want to make changes. 

To accomplish this, I've setup 3 arrays within the JS file. Each array corresponds to either Sides, Drinks, or Sauces, and within them are the IDs of the checkboxes in the DOM. Now JS has something to loop through -- it will loop through these arrays and the contents of each index will be inserted as the ID for a checkbox so that JS can determine if it's `true` or `false`. If `true`, it will retrieve the `value` of that element and push that data into the shopping cart array. Below is how the code would look:

```
for(let i = 0; i < sidesArray.Length; i++) {
     let checkBox = document.getElementById(sidesArray[i]);
     let dataValue = document.getElementById(sidesArray[i]).value;

     if(checkBox.checked == true) {
           shoppingCartArray.Sides.sides[].push(dataValue);
    }

}
```
The above loop would be performed for Drinks and Sauces as well. 

The shopping cart object array will need to be formatted something like this:

```
[
   {
      "ItemTitle": (data will be retrieved from the title element of the page),
      "Sides": sides[], <----- This name/value pair contains an array which will store selected sides
      "Drinks": drinks[],
      "Sauces": sauces[],
      "Quantity": 
      "Instructions":
      "ItemTotal": 
   }
]
```

The only other issue is transferring this data to a COOKIE so that it can retrieved on the checkout page. 

TOTAL TIME IN PROJECT SO FAR: 33 hours, 56 mins

**Comment by NullOperator404** [comment]
UPDATE 1/3/2025

A brief Google search today returned the necessary code and instructions to send order data to the POS system. Code snippet from Google provided below, along with instructions:

"To send order data from your custom website to a Clover POS, you need to utilize the Clover API, which allows you to create orders directly within your Clover system by sending a POST request with the necessary order details through your website's backend code, effectively syncing online orders with your physical POS system; this typically involves setting up a connection with your Clover merchant account, authenticating with an API key, and structuring the order data in a specific JSON format before sending it to the relevant Clover API endpoint. 

**Key steps involved:**

**Access Clover Developer Portal:**
Go to the Clover developer website to access the API documentation and obtain your merchant ID and API key needed for authentication. 

**Choose API Endpoint:**
Identify the appropriate Clover API endpoint for creating orders, usually found under "Orders" or "Atomic Orders" depending on your specific Clover version. 

**Build Order Data:**
Customer Information: Gather customer details like name, email, phone number, delivery address (if applicable). 
Order Items: List each item ordered with details like product name, quantity, price, any modifiers or options. 
Payment Information: If you are integrating with a payment gateway, you may need to send payment details, but this is often handled separately. 

**Formulate API Request:**
HTTP Method: Use a POST request to send the order data to the chosen API endpoint. 
Authorization: Include your Clover API key in the request header to authenticate. 
Payload Format: Structure the order data in a JSON format conforming to the Clover API specifications. 
Example Code Snippet (using a basic Node.js framework):"

```
const axios = require('axios');



const cloverAPIKey = 'YOUR_CLOVER_API_KEY';

const cloverMerchantId = 'YOUR_CLOVER_MERCHANT_ID';



async function sendOrderToClover(orderData) {

    const url = `https://api.clover.com/v3/merchants/${cloverMerchantId}/atomic_order/orders`;

    

    try {

        const response = await axios.post(url, orderData, {

            headers: {

                'Authorization': `Bearer ${cloverAPIKey}`,

                'Content-Type': 'application/json'

            }

        });

        console.log('Order sent to Clover successfully!', response.data);

    } catch (error) {

        console.error('Error sending order to Clover:', error);

    }

}



// Example usage:

const orderDetails = {

    "customer": { "name": "John Doe" },

    "items": [

        { "name": "Burger", "quantity": 1, "price": 10.00 }

    ]

};



sendOrderToClover(orderDetails); 
```

**Comment by NullOperator404** [comment]
UPDATE 1/5/2025

Yesterday I met with the client to go over the progress of the website and a few other issues. After seeing the Order Online screenshots, the client has asked for the Huntsville, AL location to be removed from the Order Online, so customers will only be able to order/pickup from the Butler location. This is no biggie for me, as it just means modifying the first screen on the Order Online page. 

I also learned in the meeting that their Huntsville location is not on the Clover system, so removing them from the process comes as somewhat of a relief for me, as it would've complicated the checkout process if the two locations are on two different POS systems. 

I will keep both locations on the "Locations" page for the time being, but I'm going to update that page so that the Butler location offers Online Ordering/Pickup. Both locations also need their hours of operation added.

**Comment by NullOperator404** [comment]
UPDATE 1/13/2025

Tonight I finished setting up the function for the "Add to Order" button. Clicking the button now does the following:

- Creates an empty data object to store the order data in, which will then be pushed to the shopping cart array. 
- Stores the item title to the object array
- Runs a `for` loop and determines which sides, drinks, and sauces (if any) the user has selected. If none, these values are set to `null`.
- Stores any special instructions the user inputted. 
- Stores the quantity of the item
- Stores the total price of the selected item(s)
- Pushes the completed object into the `shoppingCartArray` array.
- Converts the `shoppingCartArray` data to a string and stores it in `sessionStorage` for use on the checkout page. 

I need to run a few more tests to make sure the cart is being updated correctly (for instance, if the user adds more items, will these be added to cart & session storage correctly?). 

Now that the main functionality of the button is finished, I'm ready to work on a splash screen that will pop up with a summary of the items added. The user will be given the option to "Add more Items" or "Checkout." The "Add more Items" button will return the user to Screen 2 (the menu selection screen) where they can begin the process all over again. These selections of course, will be ADDED to the shopping cart. 

The "Checkout" button will move the user to a new page altogether. This page will summarize their cart total (only the dollar amounts; this screen will not rehash order details), will allow them to choose a pickup time (and will again give them the restaurant's address), and will contain forms for their contact info and payment info. At the bottom of this page will be a "Order & Pay Now" button which will trigger the Clover-provided code to initiate a checkout session. Order data will be passed to the Clover system and payment handled. If payment is successful, the user will be taken to a "Thank You for your Order!" screen, with perhaps a button to return them to PB's home page. 

I've "commented" the block of HTML code for the Huntsville location, as the client wished to remove this location from online media. However, the remaining location appears onscreen off-set to the left; it needs to be centered (probably just a CSS issue).

TOTAL TIME IN PROJECT SO FAR: 35 hours, 56 mins

**Comment by NullOperator404** [comment]
ADDENDUM TO PREVIOUS ENTRY:

A "Cart Total" icon needs to be added to the upper right hand side of the screen, just below the header panel. This will display current cart total (in dollars), and will display on all pages (except for checkout)

**Comment by NullOperator404** [comment]
UPDATE 1/14/2025

Preliminary HTML for the shopping cart modal is finished and I've started styling it in CSS. Tomorrow (time permitting), I will finish out the styling and start working on the JS functionality. 

TOTAL TIME IN PROJECT SO FAR: 36 hours, 31 mins

**Comment by NullOperator404** [comment]
UPDATE 1/15/2025

I performed a further test tonight to ensure additional items are being added to the `shoppingCartArray` correctly. The following procedure:

1. Select an item from the menu, selected 2 sides, and added a note to indicate this was item 1. 
2. Clicked the "Back" button to return to the menu screen, selected a different item, then two new sides, and a note indicating it was item 2. 
3. A `console.log()` run for the array indicated that both selections WERE added to the cart correctly. 

There is one minor bug the test revealed:
If the user clicks "back" and selects a different item from the menu, the previous Sides/Drinks/Sauces checkboxes will still be checked, causing a math error if the user unchecks them -- JS will subtract the dollar amount of the Side, or Drink, causing the base price for the item to be LESS than what it should be. 

SOLUTION:
To remedy the issue, a block of code needs to be added to both the "back" button function and the "Add to Cart" button which will revert all selections to "unchecked." This should be easily done by utilizing the `sides`, `drinks`, & `sauces` arrays I setup for the Add to Cart function. Just setup `for` loops that will cycle through the boxes and set each of them to `false`. (unchecked).

**Comment by NullOperator404** [comment]
UPDATE 1/17/2025

Tonight I worked on finishing the CSS styling for the cart modal. The only thing left now is to add an "Add More Items" button which will return the user to the main menu screen. I have yet to address the bug mentioned in my previous post. Now that most of the HTML & CSS is finished for the cart screen, I'll be diving into JavaScript again next to add functionality. Below is a list of JS functions that need to be added:

1. A block of code needs to be added to the `updateCart()` function that will disable the pointer events of the main screen, darken the background, and reveal the cart modal. 
2. A second block of code needs to be added to the `updateCart()` function that will populate the cart modal's item list and calculate dollar amounts to be displayed. This function will also need to calculate tip amounts to be display on-screen. 
3. A function for the "Delete item" icon that appears in each row. Just like the delete icons in my Design Lab project, these will need to 
             a). Recognize which item in the cart is being deleted
             b). Remove the necessary HTML for that row
             c). Remove the correct item from the `shoppingCartArray[]`, then "redraw" the rows based the new cart data. 
             d). Recalculate dollar amounts & tip amounts and update displays based on new data. 

I'm considering adding an object/value pair to each item in the `shoppingCartArray[]` which will provide a unique ID for the item's delete icon. I'm not sure yet though; if memory serves, the way I have this setup in the Design Lab is that an `onClick()` event handler is written inline with the HTML when the row is generated (same problem as the main menu screen -- JS can't assign event listeners to elements that don't exist yet)..but I also seem to recall this approach becoming extremely convoluted. I'd like to try and come up with a simpler approach, and assigning each delete icon a unique ID (pulled from the `shoppingCartArray[]` array) means the `onClick()` event handler only needs to get the ID of the icon being clicked and it will know which row to remove. IDs could end in numbers that correspond to the item's index in the array. Would that work? Hmm... but if the user DELETES an item, then item 3 (index 2) would be moved to index 1, meaning the Delete icon IDs no longer reference the correct row. Ugh. This is always such a fustercluck. 

4. Functionality needs to be added to the "Tip" buttons, which includes updating their appearance based on which one is selected, and updating the dollar amounts based on the tip selected. Further HTML, CSS, & JS will need to be added to handle the "Custom Tip" button. 
5. Functionality for the "Checkout" button. I've decided to go with Clover's hosted checkout rather than building a custom payment page. I looked over the documentation for building a custom payment session for Clover and it's just too complicated for the price quoted for this project. This thing needs to go live soon. So clicking the "Checkout" button will redirect to a Clover hosted checkout screen, which will import all of the cart data from my array. 
6. In conjunction with number 5, two additional webpages will need to be built - a "Success Payment" page, and a "Payment Failure" page - both URLs are needed when setting up Clover's hosted checkout page. 

Other than that, I just need to add an "Add More Items" button which will close the modal, re-enable the main screen, and return the user to the main menu screen where they can make additional selections. 

TOTAL TIME IN PROJECT SO FAR: 38 hours, 25 mins

**Comment by NullOperator404** [comment]
UPDATE 1/19/2025

Tonight I've been neck-deep in writing the JavaScript mentioned in my previous comment. 

The code disabling the background and opening the cart modal is done, as well as the code needed to populate the modal based on data in `shoppingCartArray[]`. It was necessary to add an additional object/value pair to the cart data for "Image," which reads the image `src` displayed on the item screen; this is needed for a thumbnail display in the cart. 

I've also finished the code handling the math side of things -- the Subtotal, Tax, and Total are all now calculated and displayed on-screen, and the total is also displayed in the "CHECKOUT" button text. 

I've started work on the code that will handle calculating tip amounts (based on the order total) which will then be used to update the text on the Tip buttons. NOTE TO SELF: The tip vars in the `updateCart()` function need to be moved to GLOBAL, as the function handling the tip buttons will need to access them. The tip button function will read the ID of the clicked button to determine which var it needs to access (using a `Switch` statement). It will then pass this data into a new var, and the Cart totals box will need to then be modified to display the tip amount selected and the Order Total values will need to be updated to include the tip. 

After that, I will need to work on the code for the Delete item buttons. I think I have a simpler system in mind, and I also think I've discovered why the Design Lab system is so unnecessarily convoluted. 

First, a quick aside: my mistake with the Design Lab was not setting up an object array for a "shopping cart." Instead, the final checkout screen pulls data directly from the previous Quote/Quantities screen's HTML, when it should just be reading an object array. 

Now for the simpler delete approach, and how it will work:

When the user adds an item to the cart, an object/value pair will be included in the `shoppingCartArray[]` that will be something like:

`"DeleteID": null,`

When the function starts assigning values to the object, the above object/value pair will be something like this:

`dataObj.DeleteID = "delete" + shoppingCartArray.length;`

This should assign a string value to the object that will be the word "delete," followed by a number, which is derived from the number of items in the array. So the first item would be "delete1," the second would be "delete2," and so on. This is much simpler than using a row counter to assign unique identifiers (hence why I should've used a shopping cart array in the Design Lab - it could just used the length of the cart array as it's numbers). 

When the user clicks a delete icon, the function will read the ID of the icon clicked. It can then just loop through the `shoppingCartArray[]` looking for a match to the ID it read. When it finds a match, that item can be removed from the array and it can now redraw the rows based on the updated cart. Simple and effective. 

I also think it would be wise to separate the blocks of code handling the calculations and row generation in `updateCart()` to an independent function; this way, both the Delete function and `updateCart()` can just call this section. Otherwise, all of that code will have to be copied into the delete function, and there's no point in that. 

If this approach works, I will definitely be refactoring the Design Lab's checkout process to utilize this version, as it's MUCH simpler and straightforward and doesn't rely on half a dozen integer vars being synchronized. 

TOTAL TIME IN PROJECT SO FAR: 39 hours, 42 mins

**Comment by NullOperator404** [comment]
UPDATE 1/21/2025

I recalled that my old JS textbook from college had a shopping cart project in it, so I decided to reference the book to see how deleting items form a cart is handled in that project. It turns out that JS has a built-in operator, `delete`, which can be called on JS object arrays. Not only will the selected item be deleted from the array, but `delete` does so in such a manner as to not affect the number of INDICES in the array. 

I will be implementing this version on this project, as well as refactoring the Design Lab project to use it.

**Comment by NullOperator404** [comment]
UPDATE 1/21/2025

MAJOR progress tonight. The code handling calculating, displaying, & adding tips is finished. Additionally, a block of HTML for a "Custom Tip" input section is finished and is integrated into the tip function. 

The bug mentioned a few days ago where Sides, Drinks, & Sauces checkboxes would remain checked has been fixed. 

An "Add More Items" option has been added to the shopping cart modal which when clicked closes the modal and returns the user to the main menu screen. Making additional selections successfully updates the cart and the modal. 

The "Delete Item" function has been completed, and ended up being far easier than I anticipated. The entire function amounted to a handful of lines of code:

```
//DELETE ITEM FROM CART
function deleteItem(event) {
    let itemID = event.getAttribute('id');
    //LOOP THROUGH CART AND FIND MATCH
    for(let x = 0; x < shoppingCartArray.length; x++) {
        if(itemID == "delete" + x) {
            shoppingCartArray.splice(x);
            populateModal();
        }
    }
}
```
Rather than use JS's delete operator, I decided my own method would be more efficient, as the JS delete operator sets the properties of the deleted item in the array to `null`, meaning that when JS redraws the modal, a row would be generated with all values display "null". We want the data removed from memory COMPLETELY, and the `splice()` method accomplishes this. 

Just to note how it works:
When JS generates a new item row, it assigns the row's delete icon an `ID` by combining the string "delete" with current loop number. So IDs are assigned as "delete0", "delete1", "delete2" and so on. 

The `deleteItem(event)` function reads the ID of the icon clicked and then loops through the `shoppingCartArray[]` while looking for a match to the ID it read in an `if-else` statement. If it finds a match, the current item is removed from the array and then the modal is redrawn based on the NEW array data. 

What's even better about this is even if the user deliberately or accidentally deletes ALL items from the modal, the interface will not break; the `rowContainer` section will be emptied and all dollar amounts will be zeroed. If the user then clicks "Add More Items," to start over shopping, the modal will behave as though they had just begun. 

One final note: in the above code, there is a function call for `populateModal()`. This is because as I mentioned in a previous comment, the blocks of code handling the population of the modal would need to be moved into an independent function so that both the `updateCart()` and `deleteItem()` functions could access them. 

At the moment, there is only one minor bug I need to track down: 
When the user makes a 2nd selection and that selection is one of the combos, the `updateCart()` function will ignore the first Sides' prices, as they are included. However, if the user returns to the menu and makes a third selection (another combo), the function will charge for the first two sides when it shouldn't. 

Other than that, the e-commerce section is nearing completion. Particularly with the modal; all that remains is to setup a payment page that the Checkout button will redirect to, and order data sent to the client's Clover POS system. 

Product images need to be collected, and descriptions of each item also need to be obtained from client staff (probably Denise). 

Once that's done, I will close this thread, as all other tasks pertaining to the project are unrelated to the Order Online page. 

TOTAL TIME IN PROJECT SO FAR: 42 hours, 50 mins

**Comment by NullOperator404** [comment]
UPDATE 1/21/2025 (additional 16 minutes)

The bug regarding the sides being incorrectly charged is fixed. All that was needed was a reset of the var `numSides`, which increments each time the user checks a box (but this var needs to be 0 at the start of each item screen session).

Checkboxes also reset when clicking the "Back" button. 

The application now appears to be working properly and is bug-free. I am ready to begin setting up the Checkout page where data & payment will be handed off to Clover. 

TOTAL TIME IN PROJECT SO FAR: 43 hours, 6 minutes

**Comment by NullOperator404** [comment]
UPDATE 1/23/2025

Further research tonight on the Clover APIs so I can start building the functions to accept payment & send order data to the POS. 

Based on what I've read so far, it appears that the "Checkout" button will itself simply be a redirect to the payment page. However, the "Submit Payment" button on that page will need to handle two separate functions:

1. Accepting, validating, and processing user card info
2. Formatting and sending order data to the POS. 

NOTE: Order data should ONLY send if payment is successful. 

It seems these two tasks are broken down separately in the Clover documentation, so I think the function will look something like this:

This part will handle formatting and sending order details to the POS:
```
const axios = require('axios');



const cloverAPIKey = 'YOUR_CLOVER_API_KEY';

const cloverMerchantId = 'YOUR_CLOVER_MERCHANT_ID';



async function sendOrderToClover(orderData) {

    const url = `https://api.clover.com/v3/merchants/${cloverMerchantId}/atomic_order/orders`;

    

    try {

        const response = await axios.post(url, orderData, {

            headers: {

                'Authorization': `Bearer ${cloverAPIKey}`,

                'Content-Type': 'application/json'

            }

        });

        console.log('Order sent to Clover successfully!', response.data);

    } catch (error) {

        console.error('Error sending order to Clover:', error);

    }

}



// Example usage:

const orderDetails = {

    "customer": { "name": "John Doe" },

    "items": [

        { "name": "Burger", "quantity": 1, "price": 10.00 }

    ]

};



sendOrderToClover(orderDetails); 
```

While this part will handle payment: (merchantID & token for illustration only)
```
const clover = new Clover('080391318a9041e4d122082d369b64c7', {
            merchantId: 'M1V5CF1WV0FSJ'
        });
        const elements = clover.elements();

        const styles = {
            input: {
                fontSize: '20px'
            }
        };

        const cardNumber = elements.create('CARD_NUMBER', styles);
        const cardName = elements.create('CARD_NAME', styles);
        const cardDate = elements.create('CARD_DATE', styles);
        const cardCvv = elements.create('CARD_CVV', styles);
        const cardPostalCode = elements.create('CARD_POSTAL_CODE', styles);
        const cardStreetAddress = elements.create('CARD_STREET_ADDRESS', styles);

        cardNumber.mount('#card-number');
        cardName.mount('#card-name');
        cardDate.mount('#card-date');
        cardCvv.mount('#card-cvv');
        cardPostalCode.mount('#card-postal-code');
        cardStreetAddress.mount('#card-street-address');

        const paymentReqData = {
            // Define your payment request data here
        };

        const paymentRequestButton = elements.create('PAYMENT_REQUEST_BUTTON', {
            paymentReqData
        });
        paymentRequestButton.mount('#payment-request-button');

        paymentRequestButton.addEventListener('paymentMethod', function(ev) {
            alert(JSON.stringify(ev));
        });

        paymentRequestButton.addEventListener('paymentMethodStart', function(ev) {
            console.log("Gpay in progress");
        });
```
https://docs.clover.com/dev/docs/using-the-clover-hosted-iframe

**Comment by NullOperator404** [comment]
UPDATE 1/24/2025

Tonight I started working on the "Payment page," which users will be directed to once they click the "Checkout" button on the shopping cart modal. 

This is where things get dicey; I'm now getting into the weeds of utilizing Clover's code to setup a link between the site and the restaurant's POS system. I hope to have the HTML & CSS finished in the next day or so and I'll start working on the JS. 

This part of the project will require testing using Clover's "sandbox" mode so I can ensure payments are being processed and that order data is being received by the POS. 

TOTAL TIME IN PROJECT SO FAR: 44 hours, 6 minutes

**Comment by NullOperator404** [comment]
UPDATE 1/25/2025

I spent an hour this evening working on the only other part of the payment page unrelated to Clover - importing the dollar totals from the previous page's Cart modal. 

To handle this, I setup a 2nd data object to be stored in `sessionStorage`. This object stores the final values of Subtotal, tax, total, tip, and the grand total (displayed on the checkout button). Clicking the "Checkout" button now saves the data to `sessionStorage` and redirects the user to the payment page. 

An `onLoad()` event now populates the cart totals box on the payment page by importing the data from `sessionStorage`. 

NOTE: When the user clicks the "pay" button,  a command needs to be added that will CLEAR `sessionStorage` so these items won't persist in their browser during future orders. 

I have started the preliminary JS provided by Clover to handle payment processing. I'm expecting to be running tests in the sandbox within the next few days. 

TOTAL TIME IN PROJECT SO FAR: 45 hours, 6 minutes

**Comment by NullOperator404** [comment]
UPDATE 1/26/2025

Burning the midnight oil tonight working on this, and I've made further progress. I've followed Clover's procedures to the letter and I'm not at the point where I'm ready to request a charge. However, the charge request code on Clover's site uses cURL, and I'm unfamiliar with this language. I will need to familiarize myself with it. 

The console isn't reporting any errors except that it can't connect to the Clover sandbox; it can't find a match for the merchant ID I gave it (which I retrieved from my Clover account the Clover rep setup for me). I may need to contact Jason (Clover) about this.

That said, I do feel like I'm nearing the end of this process - once I get the payment process working, I will need to setup a cURL request that sends order data to the POS system at the restaurant. 

TOTAL TIME IN PROJECT SO FAR: 46 hours, 36 minutes

**Comment by NullOperator404** [comment]
UPDATE 1/27/2025

I've been doing a little homework tonight on charge requests. It occurred to me that what the cURL segment is doing is basically just a server-to-server POST request, which can be handled with the `Fetch` API - which I'm more familiar with. 

So unless I'm mistaken, Clover's server won't care if the request is made by cURL or by `Fetch`, as long as the credentials are correct. Tonight I decided to consult ChatGPT on how to structure a simple charge request using the `Fetch` API, and this is what it replied with:

```
const chargeCreditCard = async () => {
  const url = "https://api.paymentprovider.com/v1/charge"; // Replace with actual endpoint
  const apiKey = "your-api-key"; // Replace with your API key

  const chargeData = {
    amount: 5000, // Amount in cents (e.g., $50.00)
    currency: "USD",
    source: "tok_visa", // Replace with the tokenized credit card source
    description: "Charge for a purchase",
  };

  try {
    const response = await fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "Authorization": `Bearer ${apiKey}`, // Authentication header
      },
      body: JSON.stringify(chargeData), // Convert data to JSON
    });

    if (!response.ok) {
      throw new Error(`Error: ${response.status} ${response.statusText}`);
    }

    const result = await response.json(); // Parse the response JSON
    console.log("Charge successful:", result);
  } catch (error) {
    console.error("Charge failed:", error.message);
  }
};

chargeCreditCard();
```

This makes more sense to me, as I can write this in the JS file (I literally have no idea where cURL statements go in relation to the front-end). The only real question now is, "will it work?" Tomorrow I will implement this code into the payment page's JS. I also need to contact Jason at Clover to resolve endpoint issue - I don't have a "developer" account, and this is needed to conduct tests in sandbox mode.

**Comment by NullOperator404** [comment]
UPDATE 2/2/2025

I recently spoke with Jason at Clover regarding a developer account, and he informed me that I will likely need to speak to someone directly at Clover to have an account setup, so I will be addressing that this coming week. 

In the meantime, I decided to make the best of the downtime by working on some of the other aspects of the website that I originally planned to address after the payment system was working. 

Tonight, what I thought would be an easy task has turned into a 3 1/2 hour long work session -- setting up a "cart" icon at the top of the page and having it synchronize across all pages as the user browses. 

Setting the icon itself was straightforward, but a TON of my work tonight has consisted of refactoring code to accommodate this feature. Pages outside of the Order Online page would need to pull cart data from `sessionStorage`. It then occurred to me that ALL pages will need to include the HTML describing the shopping cart modal, as clicking on the cart icon is meant to "open the cart." Since all other pages would only require code referring to the modal itself and the cart icon, I felt it best to create a new JS file that includes strictly these functions. This way, pages like Home, Menu, etc aren't loading the colossal JS file handling Order Online functionality (I mean...those functions wouldn't be needed). 

But more to the point, this approach was NECESSARY as relying on orderScript.js meant that the console was flooded with errors triggered by JS trying to assign event listeners to elements it can't find in the DOM (those elements only exist on the Order Online page). 

Further, the `populateModal()` function had to be refactored to use data from `sessionStorage`, not `shoppingCartArray[]`, as the function is needed on other pages and array contents is non-transferrable between pages. 

A few others additions were necessary to accommodate the cart now being accessible from other pages. For testing & dev purposes, I used the Home page as my guinea pig; once the code is finalized, I'll just need to copy & paste it into the other HTML pages with a call for the cartScript.js file. Here's a brief summary of how things work now:

1. When the user initially arrives on the site, clicking the cart icon will open the modal but the "Add More Items" text will read "Add Items," as the cart is empty. Clicking this text will send the user to the Order Online page. If the user returns to any page after items have been added to the cart, the text reads "Add More Items".

2. Returning to any page after adding contents to the cart will trigger a `load` event listener that calls the `updateCartIcon()` function. This function checks `sessionStorage` to see if any cart data is present, and if so, update the cart icon to indicate number if items in the cart, and redraw the modal to reflect cart contents (the modal remains hidden though; clicking the cart icon opens the modal).

3. When the Order Online page loads, a function called `checkCart()` is run that uses an `if-else` statement to determine whether data is already present in `sessionStorage`. If so, the function pushes this existing data into the `shoppingCartArray[]` so that if/when the user adds an item, the modal will reflect their existing cart contents + the new item. If the check on `sessionStorage` returns `null`, the function ignores it altogether and proceeds normally (this condition can only be true if the user has arrived at the Order Online page WITHOUT having already added items to the cart). 

This measure was necessary because the `shoppingCartArray[]` is initialized globally in the orderScript.js file, so each time the Order Online page loads, the array is empty. Without pushing `sessionStorage` data into the array upon loading, the `updateCart()` function would not remember any items the user may have added before returning to the page. 

4. The `checkCart()` function also adjusts the appearance of the Order Online page depending on whether the cart already has items in it -- if the cart is EMPTY, it likely means the user has not yet attempted to place an order, in which case the "Select Location" screen opens first. If the cart is OCCUPIED, it likely means the user is returning to the Order Online page from a cart modal on one of the other pages, perhaps to add more items. In this scenario, the page loads with the main menu screen already populated/loaded; the "select location" screen is omitted, as the user has already initiated the order process. 

For the most part, everything seems to be working correctly now (surprisingly, considering the tsunami of bugs this feature unleashed). There are still a couple of bugs that need to be addressed: 

1. When the user clicks the "Checkout" button on the modal, the "Tip" amount is showing `null`. This bug is only triggered if the user initiates an order, navigates to another page, opens the cart, then returns to the order online page to make further selections. 

2. Clicking the Delete icon no longer clears data from the modal, but I think this is because the modal is redrawn based on `sessionStorage`, not the `shoppingCartArray[]`, which is what the delete icon updates. A line of code will need to be added to the `deleteItem()` function that updates `sessionStorage`. 

TOTAL TIME IN PROJECT SO FAR: 50 hours, 8 minutes

**Comment by NullOperator404** [comment]
UPDATE 2/3/2025

Another lengthy morning tackling an endless variety of bugs. 

The first bug listed above referring to "tips" on the checkout page has been corrected. 

However, the delete command problem has turned into a hours-long session that remains unresolved. 

First, after fixing the `deleteItem()` function so that it appends `sessionStorage`, I learned that clicking delete on the FIRST item in the cart causes a cascade reaction that clears the cart entirely (and this was a problem I now remember facing in the Design Lab). 
The problem stems from element 0 being spliced from `shoppingCartArray[]`, which in turn moves elements 1, 2, 3, etc down 1 slot, meaning that what was previously element 1 is now element 0, so the `if` statement again evaluates to `true`, causing JS to delete that item. The process repeats, with the next element again occupying index 0, again resulting in a deletion. 

Likewise, this problem will manifest if the user clicks delete on any item that ISN'T the last item in the cart; clicking item 2 for example, will delete item 2 and every item after it. 

To solve this problem, I had to refactor the `deleteItem()` code to use JS's built-in `delete` operator, which assigns a value of `null` to the index instead of deleting it outright. This means that index 0 is never occupied by other items in the cart. However in doing so, I had to modify how the display for number of items in the cart is calculated. Before, I just used `shoppingCartArray.length`, but this would now display an incorrect value since deleted items still exist as an index in the array. (If item 0 of 3 is deleted, the array size is still 4). 

The two instances where cart size is calculated is within `populateModal()` and `updateCartIcon()`. In both cases, a `for` loop is setup to iterate through the cart looking for any items that are NOT equal to `null`. A var called `itemCount` increments by 1 for each time the loop finds an item not equal to `null`. `itemCount` is now used to output the correct value. Code below:

```
let itemCount = 0;
//LOOP THROUGH SESSION STORAGE TO SEE IF ANY ITEMS HAVE BEEN DELETED (VALUE NULL)
            for(let b = 0; b < parseContents.length; b++) {
                if(parseContents[b] !== null) {
                    itemCount++;
                }
            }
```

Now for the NEW PROBLEMS:

1. After exiting the cart modal by clicking "Add More Items" and then clicking the cart icon to open the modal, the deleted item has reappeared in the list. I have absolutely no idea how this is happening, as the `openCart()` function only affects VISUAL elements by modifying CSS; it does not call or implement any functions that would affect the CONTENTS of the cart. 

I will have to work on this problem later tonight or tomorrow. 

TOTAL TIME IN PROJECT SO FAR: 53 hours, 38 minutes

**Comment by NullOperator404** [comment]
ADDENDUM:

The above problem was fixed moments later. For some reason, I had an event listener setup on the cart button to call the `updateCart()` function (which redraws the modal via `populateModal()` ). Removing this event listener fixed the problem. 

Further checks by adding items to the cart, returning to the menu, opening the cart, returning to the menu, navigating to the Home page, then navigating back to the Order Online page & clicking the cart still yielded correct results.

**Comment by NullOperator404** [comment]
UPDATE 2/3/2025 10:05 PM

I wanted to try and finish working on the shopping cart issues tonight, and I think I've got everything working. I only need to copy/paste the code into the other pages, but the Cart Modal now functions the same as it does on the Order Online page, and navigating to Checkout from the modal if on the Home page still yields correct results. 

Additionally, I've added a "close" button at the top of the modal in case users want to exit the modal without having to click "Add More Items" and unwittingly transferred to the main menu screen (they may want to continue browsing after just checking their cart). 

With this completed, shopping cart functionality is finished; all that's left is to finish the server-to-server transaction phase, which will require a Clover developer account. 

Until then, I'm going to setup the Catering page and implement the Daily Specials sections - an interactive banner on the home page that will display a given day's special, a section in the Order Online main menu for the daily special (also interactive, as it will update in tandem with the home page), and of course, adding the entire Daily Specials list to the "Menu" page. Once that's finished, all that will remain is setting up an @media section in CSS for the mobile version. 

TIME IN PROJECT SO FAR: 54 hours, 38 minutes

**Comment by NullOperator404** [comment]
UPDATE 2/27/2025

Following the addition of new menu items at the restaurant, I have completed updating the code for the "Pizzas" option. Key updates are as follows:

1. A `<div>` section has been setup on the Item Screen to handle pizza topping selections. This section functions the same way as Sides & Drinks. 

2. The "Toppings" section's `display` is set to `none` by default, as this section does not need to appear unless users have selected a pizza from the main menu. In the `openOrderScreen()` function, an `if` statement checks the `id` of the item clicked to see if it's a pizza. If so, JS will set the `display` of the Toppings section to `block` to make it visible. Returning to the main menu screen either by clicking the "Back" button or by clicking "Add More Items" in the modal will reset the `display` to `none`. 

3. A new function called `selectToppings()` has been written that handles changes when the user clicks one of the Toppings checkboxes. This function works similarly to "Sides" in that it keeps track of how many toppings the user has added. On 10" and 12" 1-topping pizzas, the first topping is free, so the function will only upcharge for toppings once `numToppings` exceeds 1. 

4. Changes to the `populateModal()` function were necessary so that the selected toppings will display in the item row. 

NOTE:
One other condition needs to be addressed: the "Meat Lovers" pizza will require several modifications:

1. If the user clicks "Meat Lovers" pizza, then all meat toppings need to be set to `checked` when the Item Screen opens and `numToppings` needs to be set to 5. This will allow the price to decrease if a user then deselects one or more meat toppings. 

The client has also add a "Salads" section, so a `<div>` similar to Toppings, Sides, Drinks, & Sauces will need to be added that lists available salad dressings. 

An "Add-ons" section needs to be rebuilt, but should only appear on select items. 

TOTAL TIME IN PROJECT SO FAR: 63 hours, 48 minutes

**Comment by NullOperator404** [comment]
UPDATE 5/14/2025

Notes from ChatGTP to be used:

✅ 1. Frontend: Tokenize Card & Send Order Data
You'll collect:

Line items (items, price, quantity)

Customer details

Total amount

Tokenized card

Then send them to your backend:

js
```
const result = await clover.createToken();

if (result.token) {
  const res = await fetch('/process-payment', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      token: result.token,
      amount: 4500, // $45.00
      currency: 'USD',
      orderItems: [
        { name: 'Burger', price: 1500, quantity: 1 },
        { name: 'Fries', price: 1000, quantity: 1 },
        { name: 'Drink', price: 2000, quantity: 1 }
      ]
    }),
  });
}
```
✅ 2. Backend: Create Order in Clover POS
js
```
const CLOVER_BASE = 'https://sandbox.dev.clover.com/v3/merchants'; // use prod for live

// Helper: Create order in Clover POS
async function createCloverOrder(merchantId, token, orderItems) {
  const res = await fetch(`${CLOVER_BASE}/${merchantId}/orders`, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      lineItems: orderItems.map(item => ({
        name: item.name,
        price: item.price,
        quantity: item.quantity,
      })),
    }),
  });

  if (!res.ok) throw new Error('Failed to create Clover order');
  return res.json(); // returns { id: 'order_id', ... }
}
```
✅ 3. Backend: Charge the Card & Link to Order
js
```
app.post('/process-payment', async (req, res) => {
  const { token, amount, currency, orderItems } = req.body;
  const merchantId = process.env.CLOVER_MERCHANT_ID;
  const apiKey = process.env.CLOVER_API_KEY;

  try {
    // Step 1: Create Order in Clover
    const order = await createCloverOrder(merchantId, apiKey, orderItems);

    // Step 2: Charge the Card
    const chargeRes = await fetch('https://scl-sandbox.dev.clover.com/v1/charges', {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${apiKey}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        amount,
        currency: currency || 'USD',
        source: token,
        metadata: {
          clover_order_id: order.id,
        },
      }),
    });

    const chargeData = await chargeRes.json();
    if (!chargeRes.ok) throw new Error(chargeData.message);

    res.status(200).json({ success: true, charge: chargeData, order });
  } catch (err) {
    console.error('Payment processing error:', err);
    res.status(500).json({ success: false, message: err.message });
  }
});
```
🔐 Notes & Reminders
Use sandbox keys while testing.

You’ll need your merchant ID and a Clover API OAuth token with orders:write and payments:write scopes.

Link payments to Clover orders via the metadata field.

In production, switch from sandbox.dev.clover.com to the live api.clover.com.

**Comment by NullOperator404** [comment]
UPDATE 9/27/2025

The client has recently informed me that he intends to stick with Clover for his POS system, so I've resumed work on this project to get it finalized and published. 

I was able to open a developer account with Clover. The test merchant ID and other credentials are now working; no errors. 

I'm now in a position where I can start using Clover's test credit card info to run mock transactions. My developer account will track "sales" and indicate if any transactions are successful. 

At the moment, a `console.log()` statement seems to indicate that the form data is not being submitted when the "Pay Now" button is clicked. After reviewing the code, there seems to be a disconnect between the user clicking the pay button and the form submission code running, as if the latter isn't getting triggered. The good news is that I feel this is a relatively simple problem to resolve and I'm confident I can have the payment process working in the next few days. There's also the task of setting up and passing order info to the POS system, but should just be a matter of adding additional data to the `Fetch` request.

**Comment by NullOperator404** [comment]
UPDATE 10/6/2025

SUCCESS!

After days of troubleshooting a plethora of backend errors and navigating Clover's broken documentation, I've managed to get the charge process working. My test charge now shows as a completed sale in my developer test merchant account for Clover. 

I will run several more tests to ensure it's working properly. 

The next step will be adding order details to the request so that they can be sent to the POS system at the restaurant. 

Tonight was a major breakthrough and a huge step forward in the completion of this project. Once the order details aspect of the request is working, all that remains will be minor bugs, updates, etc to address before the site can be deployed live.

**Comment by NullOperator404** [comment]
UPDATE 10/6/2025 (1 hour later)

I've consulted ChatGPT for attaching order details. I've added the necessary code to the server.js file, but there is a minor issue on the frontend -- the way I have the shopping cart structured does not conform to what the Clover API is expecting. 

In my version, each primary item is stored in a "Item Title" key, while subsequent keys - "Sides," "Sauces,"  "Drinks," etc contain arrays listing the added items. 

However the Clover API expects each individual item to be a separate line item -- that is, sides, drinks, sauces, etc are all treated as separate items with their own individual prices. So I will need to modify my cart structure to accommodate Clover's API. The modifications will need to be as follows:

Instead of this:
```
//CREATE DATA OBJECT
    let dataObj = {
        "ItemTitle": null,
        "Sides": [],
        "Drinks": [],
        "Sauces": [],
        "Toppings": [],
        "AddOns": [],
        "Quantity": null,
        "Instructions": null,
        "ItemTotal": null,
        "Image": null
    };
```

It will need to look like this:
```
//CREATE DATA OBJECT
    let dataObj = {
      "ItemTitle": null,
      "Quantity": null,
      "Instructions": null,
      "ItemTotal": null,
      "Image": null
    };
```

The cart functions will need to be modified so that clicking an item on the main menu populates index 0, and adding sides, drinks, sauces, etc will push() additional objects into the array. `ItemTotal` will need to reflect the price of the individual item, not the order total.

**Comment by NullOperator404** [comment]
UPDATE 10/7/2025

After further brainstorming on the above mentioned problem, I've concluded that modifying the `dataObj` array would cause far too many problems. 

Foremost is the issue of the modal and checkout screen being populated -- these two screens need "Sides" and "Drinks" separate from the main item, and need to also be attached to the item they were ordered with. For example, ordering a COB with fries means the side of fries needs to be associated with the COB so that it will appear in that item's row. The Clover version does not distinguish between main item, sides, drinks, sauces, etc; each line item is treated equally. 

The other major problem would be that if the user deletes an item from the cart, the Clover version would have no way of knowing which sides, drinks, sauces, etc to remove since all items sold are simply listed in a series with no distinction among them. 

For these reasons, I feel the best approach would be to move the building of a Clover cart to the `saveTotals()` function, which fires when the user clicks the Checkout button in the modal. At this stage in the ordering process, the user has finalized their order and no further changes will be made, so cart data can now be sent to Clover. This approach keeps my existing logic handling any customization of the order as it's being built. 

However, this approach does present a few minor issues of it's own:

1. Each side, drink, etc must have a price attached to it for Clover. If the user has selected a combo, then the first 2 sides listed would need a price of 0.00, since they're included in the meal. Some sort of identifying logic will need to be written that tells JS to assign a price of 0.00 if the selected item is a combo. This shouldn't be too hard to figure out. 

2. A third key-value pair will need to be added to sessionStorage so that the Clover-friendly cart contents can be transferred to the checkout screen, where it can then be sent to Clover upon a successful charge.

**Comment by NullOperator404** [comment]
UPDATE 10/9/2025

Logic has been added to build a Clover-friendly object-array which can then be added to sessionStorage and then retrieved on the checkout page. Upon a successful transaction, the Clover "cart" will be sent to the Clover POS. Below is the code I wrote to build the Clover-friendly version, as well as what the contents looks like when logged to the console:

CODE:
```
//BUILD CLOVER-FRIENDLY CART
    let nextItem;
    let cloverCart = {
        "title": "Online Order #0001",  //INCREMENT SYSTEM WILL NEED TO BE BUILT FOR ORDER NUMS
        "state": "open",
        "total": cartTotals.FinalTotal,
        "lineItems": []
    };
    for(let c = 0; c < shoppingCartArray.length; c++) {
        //ADD MAIN ITEM
        nextItem = {
                "name": shoppingCartArray[c].ItemTitle,
                "price": 1500,
                "unitQty": 1
        }
        //PUSH NEW ITEM INTO CLOVER CART
        cloverCart.lineItems.push(nextItem);
        
        //ADD SIDES
        for(let s = 0; s < shoppingCartArray[c].Sides.length; s++) {
            nextItem = {
                "name": shoppingCartArray[c].Sides[s],
                "price": 300,
                "unitQty": 1
            }
            //PUSH NEW ITEM INTO CLOVER CART
            cloverCart.lineItems.push(nextItem);
        }
        //ADD DRINKS
        for(let d = 0; d < shoppingCartArray[c].Drinks.length; d++) {
            nextItem = {
                "name": shoppingCartArray[c].Drinks[d],
                "price": 500,
                "unitQty": 1
            }
            //PUSH NEW ITEM INTO CLOVER CART
            cloverCart.lineItems.push(nextItem);
        }
        //ADD SAUCES
        for(let g = 0; g < shoppingCartArray[c].Sauces.length; g++) {
            nextItem = {
                "name": shoppingCartArray[c].Sauces[g],
                "price": 0,
                "unitQty": 1
            }
            //PUSH NEW ITEM INTO CLOVER CART
            cloverCart.lineItems.push(nextItem);
        }
        //ADD TOPPINGS
        for(let t = 0; t < shoppingCartArray[c].Toppings.length; t++) {
            nextItem = {
                "name": shoppingCartArray[c].Toppings[t],
                "price": 150,
                "unitQty": 1
            }
            //PUSH NEW ITEM INTO CLOVER CART
            cloverCart.lineItems.push(nextItem);
        }
        //ADD ADD-ONS
        for(let a = 0; a < shoppingCartArray[c].AddOns.length; a++) {
            nextItem = {
                "name": shoppingCartArray[c].AddOns[a],
                "price": 100,
                "unitQty": 1
            }
            //PUSH NEW ITEM INTO CLOVER CART
            cloverCart.lineItems.push(nextItem);
        }

    }
```

CART OUTPUT LOOKS LIKE THIS:
```
{title: 'Online Order #0001', state: 'open', total: '41.40', lineItems: Array(10)}
lineItems
{name: 'COB-P with Cheese', price: 1500, unitQty: 1}
{name: "Smoked Mac 'n Cheese", price: 300, unitQty: 1}
{name: 'Lays Potato Chips', price: 300, unitQty: 1}
{name: 'Southern Sweet Tea', price: 500, unitQty: 1}
{name: 'Add Mustard', price: 100, unitQty: 1}
{name: 'New York Hot Dog', price: 1500, unitQty: 1}
{name: "Smoked Mac 'n Cheese", price: 300, unitQty: 1}
{name: 'Smoked Baked Beans', price: 300, unitQty: 1}
{name: 'Coca-Cola', price: 500, unitQty: 1}
{name: ' Add Cheese', price: 100, unitQty: 1}
state: "open"
title: "Online Order #0
```

Note that items are listed in the cart in their respective order -- a main item is listed first, followed by accompanying sides, drinks, sauces, add-ons, etc -- then the next main item (if others exist) are listed next. 

Additional logic will need to be added to the `openOrderScreen()` function that sets the value of a var to the base price of the item, which can later be assigned to a key in the `shoppingCartArray`. This data will then be needed to assign base prices to each main item in the Clover cart.

**Comment by NullOperator404** [comment]
UPDATE 10/15/2025

Work is virtually finished on the orders end of things -- order data is now passed to sessionStorage, retrieved on the Checkout page, and then sent to the Clover server upon successful payment. Line items are now appearing in the "Orders" section of my Clover sandbox merchant account, indicating successful transmission of order data to Clover. 

Getting this to work took a few days longer than I anticipated, as I encountered similar documentation problems with Clover just as I did the with `/charges` section. 

All that's left now is to attach the order data to the charge, and work out a minor bug: 
Clover seems to be treating each line item as an "order" then charging a third amount for the total. However, I think this is because the line items are priced pre-tax and the amount sent to Clover from the browser already has tax added. Clover offers a key-value pair which can send tax into to the server. I have since changed the value of `amount` to `subTotal` instead of `finalTotal`, which will send the pre-tax total. Clover will handle the tax on their end (though post-tax amount will still be displayed to the user on the front-end during checkout). 

Once the above steps are complete, the transaction process should be finished and I can focus on tying up loose ends on the project. 

NOTE: Before the site goes live, the following changes will need to be made:

- The `MERCHANT_ID` in all instances will need to be changed to the client's ID, which I will need to retrieve from them at the restaurant. 

- The URL endpoints will need to be changed from sandbox to production. 

- Both the e-Comm key and the orders API key will need to be changed to the client's - again retrieved from their account.

**Comment by NullOperator404** [comment]
UPDATE 10/27/2025

After further adjustments to the server.js file, I've now got the transaction process fully working -- an order is created, order data is passed to Clover, a charge is made, and the charge is attached to the order data that displays on the restaurant's POS system. A final "live" test will need to be performed, but the only things that will be changed are the Clover endpoint URLs and the `MERCHANT_ID`. 

Minor tasks need to be addressed now:

**_DONE_** 1. A numbering system needs to be setup that generates order numbers.

**_DONE_** 2. The checkout page also collects customer name, phone number, etc -- this data needs to be sent to Clover.

**_DONE_** 3. Conditional logic needs to be added to account for Combo items so that the first 2 sides & drink will not be charged at POS. 

4. Taxes need to be added at checkout, per Clover. Tax amount added by Clover must match amount indicated on front end. 

**_DONE_** 5. A redirect page needs to be added after checkout that shows customer's order number, pickup location, estimated time, etc. A button on this page can return the customer to the website's landing page. Cart data should be purged at some point. 

Once these tasks are completed, all that will remain will be tying up loose ends on the site. It's now conceivable it could be live by end of the week.

**Comment by NullOperator404** [comment]
UPDATE 11/3/2025

All of the above tasks are finished except for number 4 - the taxes. I attempted to link to Clover's `/taxRates` endpoint to retrieve a tax ID for testing, but Clover does not allow access to this endpoint from Sandbox accounts, so this aspect of the checkout will HAVE to be implemented & tested once the site is live and the backend is linked to the client's real IDs & keys. 

Since the rest of the checkout process is finished, I am closing this thread and will open a new issue for the taxes. This part of the project is now completed.
